lib/fluent/plugin/out_kafka2.rb

```diff
@@ -151,6 +153,15 @@ def configure(conf)
         if @chunk_key_tag
           log.warn "default_topic is set. Fluentd's event tag is not used for topic"
         end
+        if @create_topic
+          raise Fluent::ConfigError, "Cannot set both use_default_for_unknown_topic and create_topic"
```

Comment:
Probably it's not the code what you intended.
It should check `@use_default_for_unknown_topic` as described in the error message but the actual code checks `@default_topic.nil?`. Users cannot set `default_topic` due to this .

Reply:
FYI: When I test this patch with the following the config, it causes crash due to empty topic (I intended to forward all unmatched events to `test` topic by `default_topic` but can't do it due to above issue):

```xml
<source>
  @type forward
  port 24224
  bind 0.0.0.0
</source>

<filter test.**>
  @type record_transformer
  auto_typecast true
  enable_ruby true
  <record>
    topic ${tag}
    topic_template ${tag.split('.')[0]}
    types topic:string, topic_template:string
  </record>
</filter>
 
<match **>
  @type copy
  <store>
    @type stdout
  </store>
  <store>
    @type kafka2
    create_topic true
    topic_template_key topic_template
    brokers localhost:9092
    #default_topic test # cannot set
    <format>
      @type json
    </format>
    <buffer topic_template,topic>
      flush_interval 10s
    </buffer>
  </store>
</match>
```

Reply:
It was meant not to be used together with default topic. Use should configure only one option - to use default topic for non-existing topics or to use topic template and create topics if it does not exist.

Otherwise code should be modified to use default topic if event does not have topic template in it's key. But in my opinion it's better to always ensure topic template or topic is set than relying on default topic option.

Reply:
Pushed change that should allow it. Can you please check?

Reply:
> Pushed change that should allow it. Can you please check?

Hmm, it doesn't the one I meant.
I meant that `default_topic` should be created in advance by a user, shouldn't created automatically by fluentd.

> But in my opinion it's better to always ensure topic template or topic is set than relying on default topic option.

I agree with you in this point.

On the other hand, forcedly rejecting `default_topic` is overkill, rejecting only `use_default_for_unknown_topic` is enough. We should give a user flexibility of the setting as possible, as I showed in above config (I don't intend to create `default_topic` automatically in this config, it's created in advance.) Since `default_topic` is `nil` by default, I think it doesn't conflict with the usage you meant.

BTW `default_topic` isn't only for `use_default_for_unknown_topic` but also used at the following, do you notice it?
https://github.com/fluent/fluent-plugin-kafka/blob/6fc25e112705a3e3f50c06d1a1157296ac4708a4/lib/fluent/plugin/out_kafka2.rb#L223
I intended to forward untagged events such as Fluentd's itself log to a pre-created topic.

Reply:
Here is the patch for a1eef95861dd355257b331dec2c180eaf3d2e79e I meant
```diff
From b2cfc45947fa7413f2b10d5b7090e0fc1022d895 Mon Sep 17 00:00:00 2001
From: Takuro Ashie <ashie@clear-code.com>
Date: Fri, 20 Aug 2021 13:22:01 +0900
Subject: [PATCH] out_kafka2: Allow both create_topic & default_topic

In this case, default_topic should be created in advance, shouldn't
be created automatically.

Signed-off-by: Takuro Ashie <ashie@clear-code.com>
---
 lib/fluent/plugin/out_kafka2.rb | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/lib/fluent/plugin/out_kafka2.rb b/lib/fluent/plugin/out_kafka2.rb
index 0dad1ae..dbc80a5 100644
--- a/lib/fluent/plugin/out_kafka2.rb
+++ b/lib/fluent/plugin/out_kafka2.rb
@@ -153,9 +153,10 @@ DESC
         if @chunk_key_tag
           log.warn "default_topic is set. Fluentd's event tag is not used for topic"
         end
-        if @create_topic
-          raise Fluent::ConfigError, "Cannot set both use_default_for_unknown_topic and create_topic"
-        end
+      end
+
+      if @create_topic && @use_default_for_unknown_topic
+        raise Fluent::ConfigError, "Cannot set both use_default_for_unknown_topic and create_topic"
       end
 
       if @create_topic
-- 
2.30.2
```

(I tried to revert your second commit 328994a433ad598d0b8af81db4b0281fa12c3ba2 and add above, but I failed to push it since this pull request doesn't allowed to edit by us.)

Reply:
I think my second commit does what you suggested - if topic does not exist and topic template key is not present in event (so topic cannot be created), it does fall back to writing into default topic.